engine/classes/Elgg/UploadService.php

```diff
@@ -0,0 +1,152 @@
+	 * </code>
+	 * In case of a file input with multiple attribute, will produce an array of numeric indices [0, 1, 2]
+	 * <code>
+	 *   <input type="file" name="upload[]" multiple />
+	 * </code>
+	 * In case of a missing input element, will return an empty array
+	 * 
+	 * @param string $input_name Form input name
+	 * @return array
+	 */
+	public function getUploadedFileIndices($input_name) {
+		$files = $this->request->files;
+		if (!$files->has($input_name)) {
+			return [];
+		}
+
+		$input = $files->get($input_name);
+		if ($input instanceof UploadedFile) {
+			$input = array($input);
+		}
+		if (empty($input)) {
+			return [];
+		}
+		return array_keys((array) $input);
+	}
+
+	/**
+	 * Returns uploaded file object or an array of uploaded file objects
+	 * 
+	 * @param string $input_name Form input name
+	 * @param string $index      Optional index of the file (if the input allows multiple files)
+	 * @return UploadedFile|false
+	 */
+	public function getUploadedFile($input_name, $index = 0) {
+		$files = $this->request->files;
+		if (!$files->has($input_name)) {
+			return false;
+		}
+
+		$input = $files->get($input_name);
+		if ($input instanceof UploadedFile) {
+			$input = array($input);
+		}
+		if (empty($input)) {
+			return false;
+		}
+		return isset($input[$index]) ? $input[$index] : false;
+	}
+
+	/**
+	 * Returns contents of the uploaded file, or false if upload failed
+	 * 
+	 * @param string $input_name Form input name
+	 * @param string $index      Optional index of the file (if the input allows multiple files)
+	 * @return string|false
+	 */
+	public function readUploadedFile($input_name, $index = 0) {
+		$input = $this->getUploadedFile($input_name, $index);
+		if (!$input || !$input->isValid()) {
+			return false;
+		}
+		return file_get_contents($input->getPathname());
+	}
+
+	/**
+	 * Writes contents of the uploaded file to an instance of ElggFile
+	 *
+	 * @param ElggFile $target     Target file entity
+	 * @param string   $input_name Form input name
+	 * @param string   $index      Optional index of the file (if the input allows multiple files)
+	 * @return ElggFile
+	 */
+	public function moveUploadedFile(ElggFile $target, $input_name, $index = 0) {
+
+		$upload = $this->getUploadedFile($input_name, $index);
+		if (empty($upload) || !$upload->isValid()) {
+			return $target;
+		}
+
+		if ($target->exists()) {
+			// delete previous file
+			unlink($target->getFilenameOnFilestore());
+		}
+
+		$originalfilename = $upload->getClientOriginalName();
+		$target->originalfilename = $originalfilename;
+		if (empty($target->title)) {
+			$target->title = htmlspecialchars($target->originalfilename, ENT_QUOTES, 'UTF-8');;
+		}
+
+		$target->upload_time = time();
+		$filestorename = elgg_strtolower("file/{$target->upload_time}{$target->originalfilename}");
+		$target->setFilename($filestorename);
+
+		$target->open('write');
+		$target->write($this->readUploadedFile($input_name, $index));
```

Comment:
Replace these 3 lines with a move_uploaded_file() call. Then we don't need the `readUploadedFile` method at all.